### PR TITLE
resolved_ts: use tracked index update region read progress

### PR DESCRIPTION
**Pull request by NingLin-P**
### What problem does this PR solve?

Problem Summary:

Currently, the `safe_ts` and `apply_index` in the leader's `RegionReadProgress` are updated in different threads, although 
both of them using `fetch_max` to ensure they are increasing, in some rare conditions there may still be problems:

One case is `safe_ts` update first:
1. resolved ts worker update the `safe_ts`
2. before the `apply_index` is updated, the leader broadcast the `(safe_ts, apply_index)` item to followers, where the `apply_index` is smaller than expected
3. the follower receives the `(safe_ts, apply_index)` item with a smaller `apply_index` and serve read request before it has enough data and may get an inconsistent result

Another case is `apply_index` update first, which will cause the follower need to catch up more data to serve read request, will fix it in the next PR

### What is changed and how it works?

What's Changed:

This PR fixes the first case (`safe_ts` update first) by not directly update the `safe_ts` in resoled ts worker, and instead using `RegionReadProgress::update_safe_ts`, which will wait for the `apply_index` catch up before update `safe_ts`

### Check List 

Tests 

- Integration test

### Release note 
- No release note

### CODE REVIEW

File: components/resolved_ts/src/endpoint.rs
Comment (BusyJay) [inline components/resolved_ts/src/endpoint.rs:90]: Why not return?

File: components/resolved_ts/src/endpoint.rs
Comment (BusyJay) [inline components/resolved_ts/src/endpoint.rs:119]: Probably should show more information like region id and their actual index value.

File: components/resolved_ts/src/resolver.rs
Comment (BusyJay) [inline components/resolved_ts/src/resolver.rs:114]: How can it prevent safe_ts being updated before apply index?

Comment (NingLin-P) [reply components/resolved_ts/src/resolver.rs:114]: `update_safe_ts` will not update `safe_ts` if `apply_index` is smaller than the passed `tracked_index` and instead putting the `(tracked_index, resolved_ts)` into the `RegionReadProgress::pending_items`, and update the `safe_ts` after `apply index` is updated by `RegionReadProgress::update_applied`

Comment (BusyJay) [reply components/resolved_ts/src/resolver.rs:114]: What if `resolved_ts` is updated before `tracked_index` is updated? Like `apply_index` is updated in apply worker, but resolved ts is updated before `apply_index` is sent.

File: components/resolved_ts/src/endpoint.rs
Comment (NingLin-P) [reply components/resolved_ts/src/endpoint.rs:90]: Return an error will require to re-register the region, and I think it is unnecessary since an error (like `EpochNotMatch`) won't change the lock data that already write to kv engine at the apply thread. /cc @5kbpers

Comment (5kbpers) [reply components/resolved_ts/src/endpoint.rs:90]: After region merge some locks could be merged, so better to return an error here or detect the type of region error.

File: components/resolved_ts/src/endpoint.rs
Comment (5kbpers) [inline components/resolved_ts/src/endpoint.rs:140]: ditto

**Review by 5kbpers — APPROVED** [review]
LGTM

File: components/resolved_ts/src/endpoint.rs
Comment (BusyJay) [inline components/resolved_ts/src/endpoint.rs:84]: If you match it using `&mut self.resolver_status` then you don't need to write `ref mut` everywhere.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

**Review by BusyJay — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]
/lgtm

**Comment by NingLin-P** [comment]
/merge cancel
Wait for CI back to life

**Pull request merged**
